Plugins/MatrixPlugin/MatrixLogic.cs

```diff
@@ -0,0 +1,334 @@
+        
+
+        public int Firewall
+        {
+            get => Parse(CurrentPerson.Firewall);
+            set
+            {
+                if (CurrentPerson.Firewall != value.ToString())
+                {
+                    CurrentPerson.Firewall = value.ToString();
+                    OnPropertyChanged();
+                    OnPropertyChanged("TotalFirewall");
+                }
+            }
+        }
+
+        public int FirewallMod
+        {
+            get => Parse(CurrentPerson.ModFirewall) + (CurrentPerson.Overclocked == "Firewall" ? 1 : 0);
+            set
+            {
+                if (CurrentPerson.ModFirewall != value.ToString())
+                {
+                    CurrentPerson.ModFirewall = value.ToString();
+                    OnPropertyChanged();
+                    OnPropertyChanged("TotalFirewall");
+                }
+            }
+        }
+
+        public int TotalFirewall => Parse(CurrentPerson.Firewall) + Parse(CurrentPerson.ModFirewall) + (CurrentPerson.Overclocked == "Firewall" ? 1 : 0);
+        
+
+        public int ActionDicePool => GetTotalAttribute(CurrentAction.ActionAttribute) +
+                GetTotalSkill(CurrentAction.ActionSkill) +
+                CurrentAction.ActionModifier +
+                ActionModifier;
+        
+
+        public int DefenceDicePool => GetTotalSkill(CurrentAction.DefenceSkill) +
+                GetTotalAttribute(CurrentAction.DefenceAttribute) +
+                CurrentAction.DefenceModifier +
+                ActionModifier;
+        
+        public int ActionModifier
+        {
+            get => actionModifier;
+            set
+            {
+                actionModifier = value;
+                OnPropertyChanged();
+                OnPropertyChanged("ActionDicePool");
+                OnPropertyChanged("DefenceDicePool");
+            }
+        }
+
+        public int GetTotalMatrixAttribute(string attribute)
+        {
+            switch (attribute)
+            {
+                case "Attack":
+                    return this.TotalAttack;
+                case "Sleaze":
+                    return this.TotalSleaze;
+                case "Data Processing":
+                    return this.TotalDataProcessing;
+                case "Firewall":
+                    return this.TotalFirewall;
+            }
+            return 0;
+        }
+
+        public int GetTotalSkill(string skill)
+        {
+            if (_character.SkillsSection.GetActiveSkill(skill) != null)
+                return _character.SkillsSection.GetActiveSkill(skill).TotalBaseRating;
+            return 0;
+        }
+        public int GetTotalAttribute(string attribute)
+        {
+            if (_character.GetAttribute(attribute) != null)
+                return _character.GetAttribute(attribute).TotalValue;
```

Comment:
`SkillsSection::GetActiveSkill()` and `Character::GetAttribute()` can both be quite expensive methods, so it makes a lot more sense to either store their results for the `null` check or use null propagation to make sure they aren't called more than once in these methods.

Also, you need to be careful with how you use `SkillsSection::GetActiveSkill()` here. Skills in Chummer are meant to work off of their `DictionaryKey` property, which is what lets Chummer fetch skills with minimal overhead when a character can have lots of them. However, a skill's `DictionaryKey` isn't necessarily going to be the same as its name, so you should be cautious with how that interacts with your setup.

Reply:
Update it in https://github.com/chummer5a/chummer5a/pull/4563/commits/669e70ff69c5205c0c480d38249a1685fd174b95